File: nRF Toolbox/nRF Toolbox/BGMViewController.m
//
//  BGMViewController.m
//  nRF Toolbox
//
//  Created by Aleksander Nowakowski on 10/01/14.
//  Copyright (c) 2014 Nordic Semiconductor. All rights reserved.
//

#import "BGMViewController.h"
#import "BGMDetailsViewController.h"
#import "BGMItemCell.h"
#import "ScannerViewController.h"
#import "GlucoseReading.h"
#import "RecordAccess.h"
#import "Constants.h"
#import "CharacteristicReader.h"
#import "HelpViewController.h"

enum
{
    ACTION_REFRESH,
    ACTION_ALL_RECORDS,
    ACTION_FIRST_RECORD,
    ACTION_LAST_RECORD,
    ACTION_CLEAR,
    ACTION_DELETE_ALL,
    ACTION_CANCEL
};

@interface BGMViewController () {
    NSDateFormatter *dateFormat;
    
    CBUUID *bgmServiceUUID;
    CBUUID *bgmGlucoseMeasurementCharacteristicUUID;
    CBUUID *bgmGlucoseMeasurementContextCharacteristicUUID;
    CBUUID *bgmRecordAccessControlPointCharacteristicUUID;
    CBUUID *batteryServiceUUID;
    CBUUID *batteryLevelCharacteristicUUID;
}

/*!
 * This property is set when the device successfully connects to the peripheral. It is used to cancel the connection
 * after user press Disconnect button.
 */
@property (strong, nonatomic) CBPeripheral *connectedPeripheral;
@property (strong, nonatomic) CBCharacteristic* bgmRecordAccessControlPointCharacteristic;
@property (strong, nonatomic) NSMutableArray* readings;
@property (weak, nonatomic) IBOutlet UITableView *bgmTableView;
//@property (weak, nonatomic) IBOutlet UIBarButtonItem *actionButton;

- (IBAction)actionButtonClicked:(id)sender;

@end

@implementation BGMViewController
@synthesize bluetoothManager;
@synthesize backgroundImage;
@synthesize verticalLabel;
@synthesize battery;
@synthesize deviceName;
@synthesize connectButton;
@synthesize connectedPeripheral;
@synthesize bgmTableView;
//@synthesize actionButton;
@synthesize recordButton;
@synthesize readings;
@synthesize bgmRecordAccessControlPointCharacteristic;


-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        // Custom initialization
        readings = [NSMutableArray arrayWithCapacity:20];
        
        dateFormat = [[NSDateFormatter alloc] init];
        [dateFormat setDateFormat:@"dd.MM.yyyy, hh:mm"];
        
        bgmServiceUUID = [CBUUID UUIDWithString:bgmServiceUUIDString];
        bgmGlucoseMeasurementCharacteristicUUID = [CBUUID UUIDWithString:bgmGlucoseMeasurementCharacteristicUUIDString];
        bgmGlucoseMeasurementContextCharacteristicUUID = [CBUUID UUIDWithString:bgmGlucoseMeasurementContextCharacteristicUUIDString];
        bgmRecordAccessControlPointCharacteristicUUID = [CBUUID UUIDWithString:bgmRecordAccessControlPointCharacteristicUUIDString];
        batteryServiceUUID = [CBUUID UUIDWithString:batteryServiceUUIDString];
        batteryLevelCharacteristicUUID = [CBUUID UUIDWithString:batteryLevelCharacteristicUUIDString];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    if (is4InchesIPhone)
    {
        // 4 inches iPhone
        UIImage *image = [UIImage imageNamed:@"Background4.png"];
        [backgroundImage setImage:image];
    }
    else
    {
        // 3.5 inches iPhone
        UIImage *image = [UIImage imageNamed:@"Background35.png"];
        [backgroundImage setImage:image];
    }
    
    // Rotate the vertical label
    self.verticalLabel.transform = CGAffineTransformMakeRotation(-M_PI / 2);
    
    bgmTableView.dataSource = self;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)actionButtonClicked:(id)sender {
    UIActionSheet *actionSheet = [[UIActionSheet alloc] init];
    actionSheet.delegate = self;
    
    [actionSheet addButtonWithTitle:@"Refresh"];
    [actionSheet addButtonWithTitle:@"All"];
    [actionSheet addButtonWithTitle:@"First"];
    [actionSheet addButtonWithTitle:@"Last"];
    [actionSheet addButtonWithTitle:@"Clear"];
    [actionSheet addButtonWithTitle:@"Delete All"];
    [actionSheet addButtonWithTitle:@"Cancel"];
    [actionSheet setDestructiveButtonIndex:ACTION_DELETE_ALL];
    [actionSheet setCancelButtonIndex:ACTION_CANCEL];
    
    //[actionSheet showFromBarButtonItem:actionButton animated:YES];
    [actionSheet showInView:self.view];
}

- (IBAction)connectOrDisconnectClicked {
    if (connectedPeripheral != nil)
    {
        [bluetoothManager cancelPeripheralConnection:connectedPeripheral];
    }
}

-(BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender
{
    // The 'scan' seque will be performed only if connectedPeripheral == nil (if we are not connected already).
    return ![identifier isEqualToString:@"scan"] || connectedPeripheral == nil;
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"scan"])
    {
        // Set this contoller as scanner delegate
        ScannerViewController *controller = (ScannerViewController *)segue.destinationViewController;
        controller.filterUUID = bgmServiceUUID;
        controller.delegate = self;
    }
    else if ([segue.identifier isEqualToString:@"details"])
    {
        BGMDetailsViewController *controller = (BGMDetailsViewController *)segue.destinationViewController;
        controller.reading = [readings objectAtIndex:[bgmTableView indexPathForSelectedRow].row];
    }
    else if ([[segue identifier] isEqualToString:@"help"]) {
        HelpViewController *helpVC = [segue destinationViewController];
        helpVC.helpText =[NSString stringWithFormat:@"-BGM (BLOOD GLUCOSE MONITOR) profile allows you to connect to your Glucose sensor.\n\n-You can read the history of glucose records."];
    }

}

#pragma mark Scanner Delegate methods

-(void)centralManager:(CBCentralManager *)manager didPeripheralSelected:(CBPeripheral *)peripheral
{
    // We may not use more than one Central Manager instance. Let's just take the one returned from Scanner View Controller
    bluetoothManager = manager;
    bluetoothManager.delegate = self;
    
    // The sensor has been selected, connect to it
    peripheral.delegate = self;
    NSDictionary *options = [NSDictionary dictionaryWithObject:[NSNumber numberWithBool:YES] forKey:CBConnectPeripheralOptionNotifyOnNotificationKey];
    [bluetoothManager connectPeripheral:peripheral options:options];
}

#pragma mark Table View Datasource delegate methods

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return readings.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    BGMItemCell *cell = [tableView dequeueReusableCellWithIdentifier:@"BGMCell"];
    
    GlucoseReading* reading = [readings objectAtIndex:indexPath.row];
    cell.timestamp.text = [dateFormat stringFromDate:reading.timestamp];
    if (reading.glucoseConcentrationTypeAndLocationPresent)
    {
        cell.value.text = [NSString stringWithFormat:@"%.1f", reading.glucoseConcentration];
        cell.type.text = [reading typeAsString];
    }
    else
    {
        cell.value.text = @"-";
        cell.type.text = @"Unavailable";
    }
    
    return cell;
}

#pragma mark Action Sheet delegate methods

-(void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    // Cancel button pressed?
    if (buttonIndex == ACTION_CANCEL)
        return;
    
    RecordAccessParam param;
    NSInteger size = 0;
    BOOL clearList = YES;
    switch (buttonIndex)
    {
        case ACTION_REFRESH:
        {
            if ([readings count] > 0)
            {
                param.opCode = REPORT_STORED_RECORDS;
                param.operatorType = GREATER_THAN_OR_EQUAL;
                param.value.singleParam.filterType = SEQUENCE_NUMBER;
                GlucoseReading* reading = [readings objectAtIndex:([readings count] - 1)];
                param.value.singleParam.paramLE = CFSwapInt16HostToLittle(reading.sequenceNumber);
                size = 5;
                clearList = NO;
                break;
            } // else, obtain all records
        }
        case ACTION_ALL_RECORDS:
        {
            param.opCode = REPORT_STORED_RECORDS;
            param.operatorType = ALL_RECORDS;
            size = 2;
            break;
        }
        case ACTION_FIRST_RECORD:
        {
            param.opCode = REPORT_STORED_RECORDS;
            param.operatorType = FIRST_RECORD;
            size = 2;
            break;
        }
        case ACTION_LAST_RECORD:
        {
            param.opCode = REPORT_STORED_RECORDS;
            param.operatorType = LAST_RECORD;
            size = 2;
            break;
        }
        case ACTION_CLEAR:
        {
            // do nothing
            break;
        }
        case ACTION_DELETE_ALL:
        {
            param.opCode = DELETE_STORED_RECORDS;
            param.operatorType = ALL_RECORDS;
            size = 2;
            break;
        }
    }
    
    // Clear the current view
    if (clearList)
    {
        [readings removeAllObjects];
        [bgmTableView reloadData];
    }
    
    if (size > 0)
    {
        NSData* data = [NSData dataWithBytes:&param length:size];
        [connectedPeripheral writeValue:data forCharacteristic:bgmRecordAccessControlPointCharacteristic type:CBCharacteristicWriteWithResponse];
    }
}

#pragma mark Central Manager delegate methods

- (void)centralManagerDidUpdateState:(CBCentralManager *)central
{
    if (central.state == CBCentralManagerStatePoweredOn) {
        // TODO
    }
    else
    {
        // TODO
        NSLog(@"Bluetooth not ON");
    }
}

- (void)centralManager:(CBCentralManager *)central didConnectPeripheral:(CBPeripheral *)peripheral
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        [deviceName setText:peripheral.name];
        [connectButton setTitle:@"DISCONNECT" forState:UIControlStateNormal];
        [self enableRecordButton];
    });
    
    // Peripheral has connected. Discover required services
    connectedPeripheral = peripheral;
    [peripheral discoverServices:@[bgmServiceUUID, batteryServiceUUID]];
}

-(void)centralManager:(CBCentralManager *)central didFailToConnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error" message:@"Connecting to the peripheral failed. Try again" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
        [alert show];
        [connectButton setTitle:@"CONNECT" forState:UIControlStateNormal];
        connectedPeripheral = nil;
        [self disableRecordButton];
        [self clearUI];
    });
}

- (void)centralManager:(CBCentralManager *)central didDisconnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        [connectButton setTitle:@"CONNECT" forState:UIControlStateNormal];
        connectedPeripheral = nil;
        [self disableRecordButton];        
        [self clearUI];
    });
}

- (void) clearUI
{
    [readings removeAllObjects];
    [bgmTableView reloadData];
    
    [deviceName setText:@"DEFAULT BGM"];
    battery.tag = 0;
    [battery setTitle:@"n/a" forState:UIControlStateDisabled];
}

-(void) enableRecordButton
{
    recordButton.enabled = YES;
    [recordButton setBackgroundColor:[UIColor blackColor]];
    [recordButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
}

-(void) disableRecordButton
{
    recordButton.enabled = NO;
    [recordButton setBackgroundColor:[UIColor lightGrayColor]];
    [recordButton setTitleColor:[UIColor lightTextColor] forState:UIControlStateNormal];
}

#pragma mark Peripheral delegate methods

-(void)peripheral:(CBPeripheral *)peripheral didDiscoverServices:(NSError *)error
{
    if (error)
    {
        NSLog(@"Error discovering service: %@", [error localizedDescription]);
        [bluetoothManager cancelPeripheralConnection:connectedPeripheral];
        return;
    }
    
    for (CBService *service in peripheral.services)
    {
        // Discovers the characteristics for a given service
        if ([service.UUID isEqual:bgmServiceUUID])
        {
            [connectedPeripheral discoverCharacteristics:@[bgmGlucoseMeasurementCharacteristicUUID, bgmGlucoseMeasurementContextCharacteristicUUID, bgmRecordAccessControlPointCharacteristicUUID] forService:service];
        }
        else if ([service.UUID isEqual:batteryServiceUUID])
        {
            [connectedPeripheral discoverCharacteristics:@[batteryLevelCharacteristicUUID] forService:service];
        }
    }
}

-(void)peripheral:(CBPeripheral *)peripheral didDiscoverCharacteristicsForService:(CBService *)service error:(NSError *)error
{
    // Characteristics for one of those services has been found
    if ([service.UUID isEqual:bgmServiceUUID])
    {
        for (CBCharacteristic *characteristic in service.characteristics)
        {
            if ([characteristic.UUID isEqual:bgmGlucoseMeasurementCharacteristicUUID] ||
                [characteristic.UUID isEqual:bgmGlucoseMeasurementContextCharacteristicUUID])
            {
                // Enable notification on data characteristic
                [peripheral setNotifyValue:YES forCharacteristic:characteristic];
            }
            else if ([characteristic.UUID isEqual:bgmRecordAccessControlPointCharacteristicUUID])
            {
                bgmRecordAccessControlPointCharacteristic = characteristic;
                
                // Enable notification on data characteristic
                [peripheral setNotifyValue:YES forCharacteristic:characteristic];
            }
        }
    }
    else if ([service.UUID isEqual:batteryServiceUUID])
    {
        for (CBCharacteristic *characteristic in service.characteristics)
        {
            if ([characteristic.UUID isEqual:batteryLevelCharacteristicUUID])
            {
                // Read the current battery value
                [peripheral readValueForCharacteristic:characteristic];
                break;
            }
        }
    }
}

-(void)peripheral:(CBPeripheral *)peripheral didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error
{
    // Decode the characteristic data
    NSData *data = characteristic.value;
    uint8_t *array = (uint8_t*) data.bytes;
    
    if ([characteristic.UUID isEqual:batteryLevelCharacteristicUUID])
    {
        uint8_t batteryLevel = [CharacteristicReader readUInt8Value:&array];
        NSString* text = [[NSString alloc] initWithFormat:@"%d%%", batteryLevel];
        
        // Scanner uses other queue to send events. We must edit UI in the main queue
        dispatch_async(dispatch_get_main_queue(), ^{
            [battery setTitle:text forState:UIControlStateDisabled];
        });
        
        if (battery.tag == 0)
        {
            // If battery level notifications are available, enable them
            if (([characteristic properties] & CBCharacteristicPropertyNotify) > 0)
            {
                battery.tag = 1; // mark that we have enabled notifications
                
                // Enable notification on data characteristic
                [peripheral setNotifyValue:YES forCharacteristic:characteristic];
            }
        }
    }
    else if ([characteristic.UUID isEqual:bgmGlucoseMeasurementCharacteristicUUID])
    {
        GlucoseReading* reading = [GlucoseReading readingFromBytes:array];
        if ([readings containsObject:reading])
        {
            // If the reading has been found (the same reading has the same sequence number), replace it with the new one
            // The indexIfObjext method uses isEqual method from GlucodeReading (comparing by sequence number only)
            [readings replaceObjectAtIndex:[readings indexOfObject:reading] withObject:reading];
        }
        else
        {
            // If not, just add the new one to the array
            [readings addObject:reading];
        }
    }
    else if ([characteristic.UUID isEqual:bgmGlucoseMeasurementContextCharacteristicUUID])
    {
        //uint8_t test[] = { 0x5F, 0x00, 0x00, 0x02, 0x01, 0xF0, 0x03, 0x13, 0xF2, 0x00, 0x22, 0x03, 0x03, 0xF0, 0x01, 0xE0 };// test data
        GlucoseReadingContext* context = [GlucoseReadingContext readingContextFromBytes:array];
        // The indexIfObjext method uses isEqual method from GlucodeReadingContext (comparing with GlucoseReading by sequence number)
        NSInteger index = [readings indexOfObject:context];
        if (index != NSNotFound)
        {
            GlucoseReading* reading = [readings objectAtIndex:index];
            reading.context = context;
        }
        else
        {
            NSLog(@"Glucose Measurement with seq no %d not found", context.sequenceNumber);
        }
    }
    else if ([characteristic.UUID isEqual:bgmRecordAccessControlPointCharacteristicUUID])
    {
        RecordAccessParam* param = (RecordAccessParam*) array;
        
        dispatch_async(dispatch_get_main_queue(), ^{
        switch (param->value.response.responseCode)
        {
            case SUCCESS:
            {
                // Refresh the table
               [bgmTableView reloadData];
                break;
            }
                
            case OP_CODE_NOT_SUPPORTED:
            {
                UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error" message:@"Operation not supported" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
                [alert show];
                break;
            }
                
            case NO_RECORDS_FOUND:
            {
                UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Status" message:@"No records found" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
                [alert show];
                break;
            }
                
            case OPERATOR_NOT_SUPPORTED:
                NSLog(@"Operator not supported");
                break;
                
            case INVALID_OPERATOR:
                NSLog(@"Invalid operator");
                break;
                
            case OPERAND_NOT_SUPPORTED:
                NSLog(@"Operand not supported");
                break;
                
            case INVALID_OPERAND:
                NSLog(@"Invalid operand");
                break;
                
            default:
                NSLog(@"Response:");
                NSLog(@"Op code: %d, operator %d", param->opCode, param->operatorType);
                NSLog(@"Req Op Code: %d, response: %d", param->value.response.requestOpCode, param->value.response.responseCode);
                break;
            }
        });
    }
}


@end

File: nRF Toolbox/BGM/StandarGlucoseMonitoring/BGMViewController.swift
//
//  BGMViewController.swift
//  nRF Toolbox
//
//  Created by Mostafa Berg on 29/04/16.
//  Copyright © 2016 Nordic Semiconductor. All rights reserved.
//

import UIKit
import CoreBluetooth
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}

fileprivate func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l > r
  default:
    return rhs < lhs
  }
}


class BGMViewController: BaseViewController ,CBCentralManagerDelegate, CBPeripheralDelegate, ScannerDelegate, UITableViewDataSource, StoryboardInstantiable {
    var bluetoothManager : CBCentralManager?
    
    //MARK: - Class properties
    var connectedPeripheral                             : CBPeripheral?
    var bgmRecordAccessControlPointCharacteristic       : CBCharacteristic?
    var readings                                        : [GlucoseReading]
    var dateFormatter                                   : DateFormatter
    var bgmServiceUUID                                  : CBUUID
    var bgmGlucoseMeasurementCharacteristicUUID         : CBUUID
    var bgmGlucoseMeasurementContextCharacteristicUUID  : CBUUID
    var bgmRecordAccessControlPointCharacteristicUUID   : CBUUID
    var batteryServiceUUID                              : CBUUID
    var batteryLevelCharacteristicUUID                  : CBUUID
    
    //MARK: - ViewController outlets
    @IBOutlet weak var battery: UIButton!
    @IBOutlet weak var bgmTableView: UITableView!
    @IBOutlet weak var connectButton: UIButton!
    @IBOutlet weak var deviceName: UILabel!
    @IBOutlet weak var recordsButton: UIButton!
    @IBOutlet weak var verticalLabel: UILabel!

    @IBAction func aboutButtonTapped(_ sender: AnyObject) {
        handleAboutButtonTapped()
    }
    
    @IBAction func actionButtonTapped(_ sender: UIButton) {
        handleActionButtonTapped(from: sender)
    }
    
    @IBAction func connectionButtonTapped(_ sender: AnyObject) {
        handleConnectionButtonTapped()
    }
    
    //MARK: - UIViewController Methods
    required init(coder aDecoder: NSCoder) {
        readings = []
        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "dd.MM.yyyy, HH:mm"
        
        bgmServiceUUID                                  = CBUUID(string: ServiceIdentifiers.bgmServiceUUIDString)
        bgmGlucoseMeasurementCharacteristicUUID         = CBUUID(string: ServiceIdentifiers.bgmGlucoseMeasurementCharacteristicUUIDString)
        bgmGlucoseMeasurementContextCharacteristicUUID  = CBUUID(string: ServiceIdentifiers.bgmGlucoseMeasurementContextCharacteristicUUIDString)
        bgmRecordAccessControlPointCharacteristicUUID   = CBUUID(string: ServiceIdentifiers.bgmRecordAccessControlPointCharacteristicUUIDString)
        batteryServiceUUID                              = CBUUID(string: ServiceIdentifiers.batteryServiceUUIDString)
        batteryLevelCharacteristicUUID                  = CBUUID(string: ServiceIdentifiers.batteryLevelCharacteristicUUIDString)
        super.init(coder: aDecoder)!
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        verticalLabel.transform = CGAffineTransform(translationX: -(verticalLabel.frame.width/2) + (verticalLabel.frame.height / 2), y: 0.0).rotated(by: -.pi / 2)
        bgmTableView.dataSource = self
    }
    
    func handleActionButtonTapped(from view: UIView) {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        alert.addAction(UIAlertAction(title: "Refresh", style: .default) { _ in
            if let reading = self.readings.last {
                let nextSequence = reading.sequenceNumber + 1
                let data = Data([
                    BGMOpCode.reportStoredRecords.rawValue,
                    BGMOperator.greaterThanOrEqual.rawValue,
                    BGMFilterType.sequenceNumber.rawValue,
                    //Convert Endianess
                    UInt8(nextSequence & 0xFF),
                    UInt8(nextSequence >> 8)
                ])
                self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
            } else {
                let data = Data([BGMOpCode.reportStoredRecords.rawValue, BGMOperator.allRecords.rawValue])
                self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
            }
        })
        alert.addAction(UIAlertAction(title: "All", style: .default) { _ in
            self.readings.removeAll()
            self.bgmTableView.reloadData()
            let data = Data([BGMOpCode.reportStoredRecords.rawValue, BGMOperator.allRecords.rawValue])
            self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
        })
        alert.addAction(UIAlertAction(title: "First", style: .default) { _ in
            self.readings.removeAll()
            self.bgmTableView.reloadData()
            let data = Data([BGMOpCode.reportStoredRecords.rawValue, BGMOperator.first.rawValue])
            self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
        })
        alert.addAction(UIAlertAction(title: "Last", style: .default) { _ in
            self.readings.removeAll()
            self.bgmTableView.reloadData()
            let data = Data([BGMOpCode.reportStoredRecords.rawValue, BGMOperator.last.rawValue])
            self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
        })
        alert.addAction(UIAlertAction(title: "Clear", style: .default) { _ in
            self.readings.removeAll()
            self.bgmTableView.reloadData()
        })
        alert.addAction(UIAlertAction(title: "Delete All", style: .destructive) { _ in
            self.readings.removeAll()
            self.bgmTableView.reloadData()
            let data = Data([BGMOpCode.deleteStoredRecords.rawValue, BGMOperator.allRecords.rawValue])
            self.connectedPeripheral?.writeValue(data, for: self.bgmRecordAccessControlPointCharacteristic!, type: .withResponse)
        })
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.popoverPresentationController?.sourceView = view
        present(alert, animated: true)
    }

    func handleAboutButtonTapped() {
        showAbout(message: AppUtilities.getHelpTextForService(service: .bgm))
    }
    
    func handleConnectionButtonTapped() {
        guard let manager = bluetoothManager, let peripheral = connectedPeripheral else {
            return
        }
        manager.cancelPeripheralConnection(peripheral)
    }
    
    func clearUI() {
        readings.removeAll()
        DispatchQueue.main.async {
            self.bgmTableView.reloadData()
            self.deviceName.text = "DEFAULT_BGM"
            self.battery.tag = 0
            self.battery.setTitle("n/a", for: .disabled)
        }
    }
    
    func enableActionButton() {
        recordsButton.isEnabled = true
        recordsButton.backgroundColor = UIColor.black
        recordsButton.setTitleColor(UIColor.white, for: .normal)
    }

    func disableActionButton() {
        recordsButton.isEnabled = false
        recordsButton.backgroundColor = UIColor.lightGray
        recordsButton.setTitleColor(UIColor.lightText, for: .normal)
    }
    
    func setupNotifications() {
        if UIApplication.instancesRespond(to: #selector(UIApplication.registerUserNotificationSettings(_:))) {
            UIApplication.shared.registerUserNotificationSettings(UIUserNotificationSettings(types: [.alert, .sound], categories: nil))
        }
    }
    
    func addNotificationObservers() {
        NotificationCenter.default.addObserver(self, selector: #selector(self.applicationDidEnterBackgroundHandler),
                                                     name: UIApplication.didEnterBackgroundNotification,
                                                     object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.applicationDidBecomeActiveHandler),
                                                     name: UIApplication.didBecomeActiveNotification,
                                                     object: nil)
    }
    
    func removeNotificationObservers() {
        NotificationCenter.default.removeObserver(self, name: UIApplication.didBecomeActiveNotification,
                                                        object: nil)
        NotificationCenter.default.removeObserver(self, name: UIApplication.didEnterBackgroundNotification,
                                                        object: nil)
    }
    
    @objc func applicationDidEnterBackgroundHandler() {
        let name = connectedPeripheral?.name ?? "peripheral"
        AppUtilities.showBackgroundNotification(message: "You are still connected to \(name). It will collect data also in background.")
    }
    
    @objc func applicationDidBecomeActiveHandler(){
        UIApplication.shared.cancelAllLocalNotifications()
    }
    
    //MARK: - CBPeripheralDelegate Methods
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil else {
            print("An error occured while discovering services: \(error!.localizedDescription)")
            bluetoothManager!.cancelPeripheralConnection(peripheral)
            return
        }
        
        for aService: CBService in peripheral.services! {
            if aService.uuid.isEqual(bgmServiceUUID) {
                peripheral.discoverCharacteristics(
                    [bgmGlucoseMeasurementCharacteristicUUID, bgmGlucoseMeasurementContextCharacteristicUUID, bgmRecordAccessControlPointCharacteristicUUID],
                    for: aService)
            } else if aService.uuid.isEqual(batteryServiceUUID){
                peripheral.discoverCharacteristics([batteryLevelCharacteristicUUID], for: aService)
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard error == nil else {
            print("Error occurred while discovering characteristic: \(error!.localizedDescription)")
            bluetoothManager!.cancelPeripheralConnection(peripheral)
            return
        }
        
        if service.uuid.isEqual(bgmServiceUUID) {
            for aCharacteristic : CBCharacteristic in service.characteristics! {
                if aCharacteristic.uuid.isEqual(bgmGlucoseMeasurementCharacteristicUUID){
                    peripheral.setNotifyValue(true, for: aCharacteristic)
                } else if aCharacteristic.uuid.isEqual(bgmGlucoseMeasurementContextCharacteristicUUID) {
                    peripheral.setNotifyValue(true, for: aCharacteristic)
                } else if aCharacteristic.uuid.isEqual(bgmRecordAccessControlPointCharacteristicUUID) {
                    bgmRecordAccessControlPointCharacteristic = aCharacteristic
                    peripheral.setNotifyValue(true, for: aCharacteristic)
                }
            }
        } else if service.uuid.isEqual(batteryServiceUUID) {
            for aCharacteristic : CBCharacteristic in service.characteristics! {
                if aCharacteristic.uuid.isEqual(batteryLevelCharacteristicUUID){
                    peripheral.readValue(for: aCharacteristic)
                    break
                }
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard error == nil else {
            print("Error occurred while updating characteristic value: \(error!.localizedDescription)")
            return
        }
        
        var array = UnsafeMutablePointer<UInt8>(OpaquePointer(((characteristic.value as NSData?)?.bytes)!))
        
        if characteristic.uuid.isEqual(batteryLevelCharacteristicUUID) {
            let batteryLevel = CharacteristicReader.readUInt8Value(ptr: &array)
            let text = "\(batteryLevel)%"
            
            DispatchQueue.main.async {
                self.battery.setTitle(text, for: .disabled)
                if self.battery.tag == 0 {
                    // If battery level notifications are available, enable them
                    if characteristic.properties.contains(.notify)
                    {
                        self.battery.tag = 1; // mark that we have enabled notifications
                        peripheral.setNotifyValue(true, for: characteristic)
                    }
                }
            }
            
        } else if characteristic.uuid.isEqual(bgmGlucoseMeasurementCharacteristicUUID) {
            print("New glucose reading")
            let reading = GlucoseReading(array)
            
            if let index = readings.firstIndex(of: reading) {
                readings[index] = reading
            } else {
                readings.append(reading)
            }
        } else if characteristic.uuid.isEqual(bgmGlucoseMeasurementContextCharacteristicUUID) {
            let context = GlucoseReadingContext(array)
            
            if let index = readings.firstIndex(where: { $0.sequenceNumber == context.sequenceNumber }) {
                readings[index].context = context
            } else {
                print("Glucose measurement with sequence number: \(context.sequenceNumber) not found")
            }
        } else if characteristic.uuid.isEqual(bgmRecordAccessControlPointCharacteristicUUID) {
            print("OpCode: \(array[0]), Operator: \(array[2])")
            DispatchQueue.main.async {
                switch BGMResponseCode(rawValue:array[2])! {
                case .success:
                    self.bgmTableView.reloadData()
                case .opCodeNotSupported:
                    AppUtilities.showAlert(title: "Error", andMessage: "Operation not supported", from: self)
                case .noRecordsFound:
                    AppUtilities.showAlert(title: "Error", andMessage: "No records found", from: self)
                case .operatorNotSupported:
                    AppUtilities.showAlert(title: "Error", andMessage: "Operator not supported", from: self)
                case .invalidOperator:
                    AppUtilities.showAlert(title: "Error", andMessage: "Invalid operator", from: self)
                case .operandNotSupported:
                    AppUtilities.showAlert(title: "Error", andMessage: "Operand not supported", from: self)
                case .invalidOperand:
                    AppUtilities.showAlert(title: "Error", andMessage: "Invalid operand", from: self)
                case .abortUnsuccessful:
                    AppUtilities.showAlert(title: "Error", andMessage: "Abort unsuccessful", from: self)
                case .procedureNotCompleted:
                    AppUtilities.showAlert(title: "Error", andMessage: "Procedure not completed", from: self)
                case .reserved:
                    break
                }
            }
        }
    }
    //MARK: - CBCentralManagerDelegate Methdos
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state == .poweredOff {
            print("Bluetooth powered off")
        } else {
            print("Bluetooth powered on")
        }
    }
    
    func centralManagerDidSelectPeripheral(withManager aManager: CBCentralManager, andPeripheral aPeripheral: CBPeripheral) {
        connectedPeripheral = aPeripheral
        connectedPeripheral?.delegate = self
        bluetoothManager = aManager
        bluetoothManager?.delegate = self
        let options = NSDictionary(object: NSNumber(value: true as Bool), forKey: CBConnectPeripheralOptionNotifyOnNotificationKey as NSCopying)
        bluetoothManager?.connect(aPeripheral, options: options as? [String : AnyObject])
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        connectedPeripheral = peripheral
        peripheral.discoverServices([bgmServiceUUID, batteryServiceUUID])
        DispatchQueue.main.async {
            self.deviceName.text = peripheral.name
            self.connectButton.setTitle("DISCONNECT", for: .normal)
            self.enableActionButton()
            self.setupNotifications()
        }
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        DispatchQueue.main.async {
            AppUtilities.showAlert(title: "Error", andMessage: "Connecting to peripheral failed. Please Try again", from: self)
            self.connectButton.setTitle("CONNECT", for: .normal)
            self.connectedPeripheral = nil
            self.disableActionButton()
            self.clearUI()
        }
    }
    
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        DispatchQueue.main.async { 
            self.connectButton.setTitle("CONNECT", for: .normal)
            self.connectedPeripheral = nil
            
            if AppUtilities.isApplicationInactive() == true {
                let name = peripheral.name ?? "Peripheral"
                AppUtilities.showBackgroundNotification(message: "\(name) is disconnected.")
            }
            self.disableActionButton()
            self.clearUI()
            self.removeNotificationObservers()
        }
    }
    
    //MARK: - UITableViewDataSoruce methods
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return readings.count
    }
    
    //MARK: - UITableViewDelegate methods
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "BGMCell") as! BGMItemCell
        
        let reading = readings[indexPath.row]
        cell.timestamp.text = dateFormatter.string(from: reading.timestamp as Date)
        
        if reading.glucoseConcentrationTypeAndLocationPresent {
            cell.type.text = "\(reading.type!)"

            switch reading.unit! {
            case .mol_L:
                cell.value.text = String(format: "%.1f", reading.glucoseConcentration! * 1000)   // mol/l -> mmol/l conversion
                cell.unit.text = "mmol/l"
                break
            case .kg_L:
                cell.value.text = String(format: "%0f", reading.glucoseConcentration! * 100000)  // kg/l -> mg/dl conversion
                cell.unit.text = "mg/dl"
                break
            }
        } else {
            cell.value.text = "-"
            cell.type.text = "Unavailable"
            cell.unit.text = ""
        }
        
        return cell
    }
    
    //MARK: - Segue methods
    override func shouldPerformSegue(withIdentifier identifier: String, sender: Any?) -> Bool {
        return identifier != "scan" || connectedPeripheral == nil
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "scan" {
            let navigationController = segue.destination as! UINavigationController
            let controller = navigationController.children.first as! ScannerViewController
            controller.filterUUID = bgmServiceUUID
            controller.delegate = self
        } else if segue.identifier == "details" {
            let controller = segue.destination as! BGMDetailsViewController
            controller.reading = readings[bgmTableView.indexPathForSelectedRow!.row]
        }
    }
}
